### ISSUE TITLE
Change indentation

### ISSUE THREAD

**Issue by 89z** [MJKWoolnough/javascript#2]
Using this program:

~~~go
package main

import (
   "fmt"
   "vimagination.zapto.org/javascript"
   "vimagination.zapto.org/javascript/walk"
   "vimagination.zapto.org/parser"
)

type visit struct{}

func (visit) Handle(t javascript.Type) error {
   fmt.Println(t)
   return nil
}

func main() {
   t := parser.NewStringTokeniser("var x = 'lorem ipsum';")
   m, err := javascript.ParseModule(t)
   if err != nil {
      panic(err)
   }
   var v visit
   walk.Walk(m, v)
}
~~~

I get this result:

~~~
ModuleItem{
	StatementListItem: StatementListItem{
		Statement: Statement{
			Type: StatementNormal
			VariableStatement: VariableStatement{
				VariableDeclarationList: [
					0:VariableDeclaration{
						BindingIdentifier: Type: Identifier - Data: "x"
						Initializer: AssignmentExpression{
							ConditionalExpression: ConditionalExpression{
								LogicalORExpression: LogicalORExpression{
									LogicalANDExpression: LogicalANDExpression{
										BitwiseORExpression: BitwiseORExpression{
											BitwiseXORExpression: BitwiseXORExpression{
												BitwiseANDExpression: BitwiseANDExpression{
													EqualityExpression: EqualityExpression{
														EqualityOperator: 
														RelationalExpression: RelationalExpression{
															RelationshipOperator: 
															ShiftExpression: ShiftExpression{
																ShiftOperator: 
																AdditiveExpression: AdditiveExpression{
																	AdditiveOperator: 
																	MultiplicativeExpression: MultiplicativeExpression{
																		MultiplicativeOperator: 
																		ExponentiationExpression: ExponentiationExpression{
																			UnaryExpression: UnaryExpression{
																				UpdateExpression: UpdateExpression{
																					LeftHandSideExpression: LeftHandSideExpression{
																						NewExpression: NewExpression{
																							News: 0
																							MemberExpression: MemberExpression{
																								PrimaryExpression: PrimaryExpression{
																									Literal: Type: StringLiteral - Data: "'lorem ipsum'"
																								}
																							}
																						}
																					}
																					UpdateOperator: 
																				}
																			}
																		}
																	}
																}
															}
														}
													}
												}
											}
										}
									}
								}
							}
							AssignmentOperator: 
						}
					}
				]
			}
		}
	}
}
~~~

On a terminal, even full screen, this is unreadable. Would it be possible to reduce, or eliminate the indentation? Compare with another package:

~~~go
package main

import (
   "fmt"
   "github.com/tdewolff/parse/v2"
   "github.com/tdewolff/parse/v2/js"
)

type visit struct{}

func (visit) Exit(js.INode) {}

func (v visit) Enter(n js.INode) js.IVisitor {
   fmt.Println(n)
   return v
}

func main() {
   ast, err := js.Parse(parse.NewInputString("var x = 'lorem ipsum';"))
   if err != nil {
      panic(err)
   }
   var vis visit
   js.Walk(vis, ast)
}
~~~

Result:

~~~
Decl(var Binding(x = 'lorem ipsum'))
Stmt({ Decl(var Binding(x = 'lorem ipsum')) })
Decl(var Binding(x = 'lorem ipsum'))
Binding(x = 'lorem ipsum')
x
'lorem ipsum'
~~~

**Comment by MJKWoolnough** [comment]
Assuming that your terminal doesn't allow you to alter the size of a tab, you can use something like the following to wrap os.Stdout and convert the indents to spaces:

```
type indent2Space struct {
	io.Writer
	pastIndent bool
}

var space = []byte{' '}

func (is *indent2Space) Write(p []byte) (int, error) {
	c := 0
	for len(p) > 0 {
		if !is.pastIndent {
			if p[0] == '	' {
				p = p[1:]
				d, err := is.Writer.Write(space)
				c += d
				if err != nil {
					return c, err
				}
			} else {
				is.pastIndent = true
			}
		} else {
			for n, b := range p {
				if b == '\n' {
					d, err := is.Writer.Write(p[:n+1])
					c += d
					if err != nil {
						return c, err
					}
					p = p[n+1:]
					is.pastIndent = false
					break
				}
			}
		}
	}
	return c, nil
}
```

It should also be noted that that Walk function doesn't automatically descend through the layers, that's left to the user. You'd need to have the Handle method of visit call walk.Walk(t, visit{}) in order to recursively go through the AST, though I don't recommend doing that with a simple print method.

**Comment by 89z** [comment]
> Assuming that your terminal doesn't allow you to alter the size of a tab, you can use something like the following to wrap os.Stdout and convert the indents to spaces:

Ideally, I dont want indentation at all. I would like all the elements to be left justified, simlar to the other package, or similar to what youd get from printing tokens with a lexer. Is that possible?

> It should also be noted that that Walk function doesn't automatically descend through the layers, that's left to the user. You'd need to have the Handle method of visit call walk.Walk(t, visit{}) in order to recursively go through the AST, though I don't recommend doing that with a simple print method.

Im confused, it seems with the extremely nesting already printed, that recursion is already being done?

**Comment by MJKWoolnough** [comment]
> > Assuming that your terminal doesn't allow you to alter the size of a tab, you can use something like the following to wrap os.Stdout and convert the indents to spaces:
> 
> Ideally, I dont want indentation at all. I would like all the elements to be left justified, simlar to the other package, or similar to what youd get from printing tokens with a lexer. Is that possible?

If all you want is the tokens then you can use something like the following to print them out:

```
package main

import (
	"fmt"
	"os"

	"vimagination.zapto.org/javascript"
	"vimagination.zapto.org/parser"
)

func main() {
	t := parser.NewStringTokeniser("var x = 'lorem ipsum';")
	javascript.SetTokeniser(&t)
	for {
		tk, err := t.GetToken()
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
		}
		if tk.Type == parser.TokenDone {
			break
		}
		fmt.Println(tk)
	}
}
```


> > It should also be noted that that Walk function doesn't automatically descend through the layers, that's left to the user. You'd need to have the Handle method of visit call walk.Walk(t, visit{}) in order to recursively go through the AST, though I don't recommend doing that with a simple print method.
> 
> Im confused, it seems with the extreme nesting already printed, that recursion is already being done?

Printing the parsed module 'm' would result in the same output, without using Walk at all. Printing a node of the AST prints itself and all its included AST nodes.

**Comment by 89z** [comment]
> If all you want is the tokens then you can use something like the following to print them out:

No, I am not wanting a tokeniser, just a parser with a similar output. What I am trying to do, is given some arbitrary JavaScript input, extract all nodes that are valid JSON. So given this input:

~~~
var x = 'lorem ipsum';
~~~

the output would just be the string `lorem ipsum`, or given this input:

~~~
var x = 'lorem ipsum';
var y = 999;
~~~

the output would be something like one of these:

~~~go
[]interface{}{"lorem ipsum", 999}

[]string{`"lorem ipsum"`, "999"}

[][]byte{
   []byte(`"lorem ipsum"`), []byte("999"),
}
~~~

**Comment by MJKWoolnough** [comment]
Okay, something like this'll be what you want:

```
package main

import (
	"fmt"

	"vimagination.zapto.org/javascript"
	"vimagination.zapto.org/javascript/walk"
	"vimagination.zapto.org/parser"
)

func jsonArray(al *javascript.ArrayLiteral) bool {
	if al.SpreadElement != nil {
		return false
	}
	for _, ae := range al.ElementList {
		if !jsonAE(&ae) {
			return false
		}
	}
	return true
}

func jsonObject(ol *javascript.ObjectLiteral) bool {
	for _, pd := range ol.PropertyDefinitionList {
		if pd.MethodDefinition != nil || pd.PropertyName == nil || pd.PropertyName.LiteralPropertyName == nil || pd.AssignmentExpression == nil {
			return false
		}
		if !jsonAE(pd.AssignmentExpression) {
			return false
		}
	}
	return true
}

func jsonAE(ae *javascript.AssignmentExpression) bool {
	if ae.AssignmentOperator != javascript.AssignmentNone || ae.ConditionalExpression == nil {
		return false
	}
	uc := javascript.UnwrapConditional(ae.ConditionalExpression)
	switch uc := uc.(type) {
	case *javascript.PrimaryExpression:
		if uc.Literal == nil {
			return false
		}
	case *javascript.ArrayLiteral:
		if !jsonArray(uc) {
			return false
		}
	case *javascript.ObjectLiteral:
		if !jsonObject(uc) {
			return false
		}
	default:
		return false
	}
	return true
}

type JSON []string

func (j *JSON) Handle(t javascript.Type) error {
	switch t := t.(type) {
	case *javascript.PrimaryExpression:
		if t.Literal != nil {
			*j = append(*j, t.Literal.Data)
		}
	case *javascript.ArrayLiteral:
		if jsonArray(t) {
			*j = append(*j, fmt.Sprintf("%s", t))
			return nil
		}
	case *javascript.ObjectLiteral:
		if jsonObject(t) {
			*j = append(*j, fmt.Sprintf("%s", t))
			return nil
		}
	}
	walk.Walk(t, j)
	return nil
}

func main() {
	m, _ := javascript.ParseModule(parser.NewStringTokeniser(`
var x = 'lorem ipsum';
var y = 999;
var z = [1, 2, "", null];
var c = {a: 1, b: 2};
var badA = [() => {}];
var badO = {a: () => {}, b: 1};
var badU = [undefined];
	`))
	var j JSON
	walk.Walk(m, &j)
	fmt.Println(j)
}
```
That should pull out all valid JSON Literals, Object and Arrays, though you should obviously do a little more testing than I've done there.

**Comment by 89z** [comment]
> Okay, something like this'll be what you want:

Thats amazing, I think thats exactly what I want! It also seems to catch internal nodes, like this:

~~~js
hello(999, 'world');
~~~

> Assuming that your terminal doesn't allow you to alter the size of a tab, you can use something like the following to wrap os.Stdout and convert the indents to spaces:

How would this fit in with the code I originally put? I tried replacing my `visit` type, but I cant seem to get any output. What type of your package calls the Write method?

**Comment by MJKWoolnough** [comment]
> > Assuming that your terminal doesn't allow you to alter the size of a tab, you can use something like the following to wrap os.Stdout and convert the indents to spaces:
> 
> How would this fit in with the code I originally put? I tried replacing my `visit` type, but I cant seem to get any output. What type of your package calls the Write method?

The easiest way to use indent2Space, would be to create a global var wrapping Stdout, like so:

```
var noIndents = &indent2Space{Writer: os.Stdout}
```

After that, just replace fmt.Println(t) with fmt.Fprintln(noIndents, t) to use the wrapped Writer.

**Comment by 89z** [comment]
> After that, just replace fmt.Println(t) with fmt.Fprintln(noIndents, t) to use the wrapped Writer.

Hm OK. That seems like a good general solution. Are you open to having option for customizing output? For example, with the other package, this code:

~~~go
fmt.Println(n)
fmt.Printf("%#v\n", n)
~~~

gives different outputs:

~~~
'lorem ipsum'
&js.LiteralExpr{TokenType:0x5, Data:[]uint8{0x27, 0x6c, 0x6f, 0x72, 0x65, 0x6d, 0x20, 0x69, 0x70, 0x73, 0x75, 0x6d, 0x27}}
~~~

while the same code with this package, gives the same output regardless. It would be nice to have an in-package way to customize the output, whether thats changing the indentation, or getting the "debug" output with `%#v`.

**Comment by MJKWoolnough** [comment]
> It would be nice to have an in-package way to customize the output, whether thats changing the indentation, or getting the "debug" output with `%#v`.

For printing, %v will print what you had above, %+v will print a more verbose form, including all of the nil values, which is pretty useful when debugging. %s will print the source and %+s will print a slightly more verbose version of the source, which usually just includes keeping original line-structure in a few places, though it does a few other things as well, such as always printing object keys.

**Comment by 89z** [comment]
Is it possible to get the type in the output though? This is really useful, as `javascript.Type` is an interface, so if someone wants to do a type assertion, they need to know what the underlying type is. Normally you can get this with `%#v`.

**Comment by MJKWoolnough** [comment]
> Is it possible to get the type in the output though? This is really useful, as `javascript.Type` is an interface, so if someone wants to do a type assertion, they need to know what the underlying type is. Normally you can get this with `%#v`.

The type is the first thing printed with %v, though it doesn't have the module name prefixed on it.

**Comment by 89z** [comment]
I appreciate all the comments, but I just cant seem to get the output in a reasonable format. I tried this code:

~~~go
package main

import (
   "fmt"
   "vimagination.zapto.org/javascript"
   "vimagination.zapto.org/javascript/walk"
   "vimagination.zapto.org/parser"
)

type visit struct{}

func (v visit) Handle(t javascript.Type) error {
   fmt.Printf("%s\n", t)
   walk.Walk(t, v)
   return nil
}

func main() {
   t := parser.NewStringTokeniser("var x = 'lorem ipsum';")
   m, err := javascript.ParseModule(t)
   if err != nil {
      panic(err)
   }
   var v visit
   walk.Walk(m, v)
}
~~~

but I get this:

~~~js
var x = 'lorem ipsum';
var x = 'lorem ipsum';
var x = 'lorem ipsum';
var x = 'lorem ipsum';
x = 'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
'lorem ipsum'
~~~

I also tried `%+v`, `%#v` and `%+s`, but they all either have the crazy indent, or missing type information.

**Comment by MJKWoolnough** [comment]
> I also tried `%+v`, `%#v` and `%+s`, but they all either have the crazy indent, or missing type information.

The fmt package does have the %T verb to get the name of a type. If that's not useful to you, can you give an example of the output you're wanting?

**Comment by 89z** [comment]
> If that's not useful to you, can you give an example of the output you're wanting?

For example with this code:

~~~go
package main

import (
   "fmt"
   "github.com/tdewolff/parse/v2"
   "github.com/tdewolff/parse/v2/js"
)

type visit struct{}

func (visit) Exit(js.INode) {}

func (v visit) Enter(n js.INode) js.IVisitor {
   fmt.Printf("%#v\n", n)
   return v
}

func main() {
   ast, err := js.Parse(parse.NewInputString("var x = 'lorem ipsum';"))
   if err != nil {
      panic(err)
   }
   var vis visit
   js.Walk(vis, ast)
}
~~~

I get this output:

~~~go
&js.AST{Comments:[][]uint8(nil), BlockStmt:js.BlockStmt{List:[]js.IStmt{(*js.VarDecl)(0xc00004c3c0)}, Scope:js.Scope{Parent:(*js.Scope)(nil), Func:(*js.Scope)(0xc00001e0d8), Declared:js.VarArray{(*js.Var)(0xc0000765a0)}, Undeclared:js.VarArray{}, NumVarDecls:0x1, NumForInit:0x0, NumArguments:0x0, IsGlobalOrFunc:false, HasWith:false}}}
&js.BlockStmt{List:[]js.IStmt{(*js.VarDecl)(0xc00004c3c0)}, Scope:js.Scope{Parent:(*js.Scope)(nil), Func:(*js.Scope)(0xc00001e0d8), Declared:js.VarArray{(*js.Var)(0xc0000765a0)}, Undeclared:js.VarArray{}, NumVarDecls:0x1, NumForInit:0x0, NumArguments:0x0, IsGlobalOrFunc:false, HasWith:false}}
&js.VarDecl{TokenType:0x823, List:[]js.BindingElement{js.BindingElement{Binding:(*js.Var)(0xc0000765a0), Default:(*js.LiteralExpr)(0xc00004c3e0)}}}
&js.BindingElement{Binding:(*js.Var)(0xc0000765a0), Default:(*js.LiteralExpr)(0xc00004c3e0)}
&js.Var{Data:[]uint8{0x78}, Link:(*js.Var)(nil), Uses:0x1, Decl:0x1}
&js.LiteralExpr{TokenType:0x5, Data:[]uint8{0x27, 0x6c, 0x6f, 0x72, 0x65, 0x6d, 0x20, 0x69, 0x70, 0x73, 0x75, 0x6d, 0x27}}
~~~

For someone just starting out with the package, this is perfect. No indentation, and types are given for each node. That way someone can easily see at a glance what is going on, and it will help to develop more complex code for people importing the package.

> The fmt package does have the %T verb to get the name of a type.

I guess I could do something like `%T %s`, but that seems a little awkward, when with normal types you can just do `%#v`. So I guess the issue is, that since the package has hijacked printf, people using it cant rely on their previous knowledge of how to do certain tasks, but have to relearn how to get the same output using this packages idioms.

**Comment by MJKWoolnough** [comment]
> For someone just starting out with the package, this is perfect. No indentation, and types are given for each node. That way someone can easily see at a glance what is going on, and it will help to develop more complex code for people importing the package.

It's funny you say that as the reason that it prints the AST the way it does was so that 'someone can easily see at a glance what is going on', letting you see a nicely formatted AST structure.

> > The fmt package does have the %T verb to get the name of a type.
> 
> I guess I could do something like `%T %s`, but that seems a little awkward, when with normal types you can just do `%#v`. So I guess the issue is, that since the package has hijacked printf, people using it cant rely on their previous knowledge of how to do certain tasks, but have to relearn how to get the same output using this packages idioms.

I mean, essentially…yes? I wrote it the way it is because without the recursion you don't really get any picture of what is going on, especially in more complex examples. Once you start looking at trees for branching code, with functions, classes, etc. the method of printing that you're suggested would become all but useless; you wouldn't know how one node relates to another.

For better or worse, the currently printing method for the AST shows you the whole tree of information. It may look like a lot of information, and it is, but I cannot think of a better way of representing a tree of data in text form. It shows the type name, it's fields and their values, recursively, ignoring any empty values unless you specify the + flag.

**Comment by 89z** [comment]
> For better or worse, the currently printing method for the AST shows you the whole tree of information. It may look like a lot of information, and it is, but I cannot think of a better way of representing a tree of data in text form. It shows the type name, it's fields and their values, recursively, ignoring any empty values unless you specify the + flag.

What youve chosen is useful, but it seems no option is available to choose the ident. For example with JSON, you have MarshalIndent [1] and SetIndent [2], where the user can choose tabs, spaces, or nothing. But with this package, the identation has been pre-chosen, and the user cant do anything about it, other than using an out of package method, like doing string replacement on the output.

Its not a huge deal, as once people are familiar with the package, they probably wont be doing much AST printing anyway, but I think having some in-package way to configure it would be very helpful.

1. https://godocs.io/encoding/json#MarshalIndent
2. https://godocs.io/encoding/json#Encoder.SetIndent

**Comment by MJKWoolnough** [comment]
> Its not a huge deal, as once people are familiar with the package, they probably wont be doing much AST printing anyway, but I think having some in-package way to configure it would be very helpful.

Unfortunately, the only way to do that would be to have a global (or, at least, package level) var to control the indentation, which is both messy and race-y. Failing that, I'd have to rewrite the printing code to not use the fmt.Formatter interface, which means you would lose the benefit of simply being able to fmt.Println a node and get the tree printed; I'm not sure that's a good trade just to control the indentation.

Also, the benefit of tabs is that they're easy to replace/resize; any decent terminal/text-editor will allow you to choose whatever size you want for them, which mitigates pretty much the entire issue here.

Having said all of that, if you still need to change the indent for learning purposes, feel free to fork the package and change the indent string, it's at the top of the [format.go](https://github.com/MJKWoolnough/javascript/blob/0623918429caa2ddf8608839106a067e9c7dc384/format.go#L14) file, a var named indent.

**Comment by 89z** [comment]
I guess Im just confused, as to why its not already possible. For example this code:

~~~go
package main
import "fmt"

type a struct {
   b int
   c string
}

func main() {
   d := a{9, "hello"}
   fmt.Println(d)
   fmt.Printf("%+v\n", d)
   fmt.Printf("%#v\n", d)
}
~~~

You get this result:

~~~go
{9 hello}
{b:9 c:hello}
main.a{b:9, c:"hello"}
~~~

So youre getting three different outputs, depending on the format string. However with this package, both `%v` and `%#v` give the same exact output it seems. Why is that, and can it be changed so that `%#v` doesnt do any indentation, but just gives the raw debug format?

**Comment by MJKWoolnough** [comment]
> However with this package, both `%v` and `%#v` give the same exact output it seems. Why is that, and can it be changed so that `%#v` doesnt do any indentation, but just gives the raw debug format?

The [fmt.Formatter interface](https://pkg.go.dev/fmt#Formatter) overrides how the fmt package formats values. Unfortunately there's no way to state that a particular verb isn't overridden and should be left to fmt to deal with. Essentially, for %#v to work, it would need to be re-implemented in this package, which is do-able, but it's work, and work I'm not sure has any real benefit.

**Comment by 89z** [comment]
> Unfortunately there's no way to state that a particular verb isn't overridden and should be left to fmt to deal with. Essentially, for %#v to work, it would need to be re-implemented in this package, which is do-able, but it's work, and work I'm not sure has any real benefit.

Well, thats a bummer. I think the other package [1] dealt with this, by having both `String` and `JS` methods, which would be similar to your `%v` and `%+v`. But by using methods instead of implementing `fmt.Formatter`, the default formatter isnt clobbered, which means you are still able to utilise `%#v`, as I have shown in my examples.

I think this is a superior approach, as you end up with the same control, in that you can add as many format methods as you like, without losing any defaults formats that you might gets from `%+v` or `%#v` etc. I can understand if you dont want to make a change like that, as it might be a big change. But maybe something to think about. Thanks again for all the comments.

1. https://godocs.io/github.com/tdewolff/parse/v2/js

**Comment by 89z** [comment]
FYI I figured out a workaround for this. You can use the Formatter interface as youve done, but still keep some default formats if you want to. Here is some example code:

~~~go
package main
import "fmt"

type hello struct {
   alfa int64
   bravo string
}

func (h hello) Format(f fmt.State, verb rune) {
   type _hello hello
   if f.Flag('#') {
      fmt.Fprintf(f, "%#v", _hello(h))
   } else if f.Flag('+') {
      fmt.Fprintf(f, "%+v", _hello(h))
   } else {
      fmt.Fprint(f, h.bravo)
   }
}

func main() {
   hi := hello{9, "nine"}
   fmt.Printf("%#v\n", hi) // main._hello{alfa:9, bravo:"nine"}
   fmt.Printf("%+v\n", hi) // {alfa:9 bravo:nine}
   fmt.Printf("%v\n", hi) // nine
}
~~~

**Comment by MJKWoolnough** [comment]
Neat hack.

Implemented as per [v0.2.4](https://github.com/MJKWoolnough/javascript/tree/v0.2.4).
